src/toil/common.py

```diff
@@ -565,6 +565,62 @@ def addGroup(headingString, bodyString):
                            "argparse.ArgumentParser" % parser.__class__)
 
 
+def getNodeID(extraIDFiles=[]):
+    """Return unique ID of the current node (host).
+    Tries several methods until success.
+    The returned ID should be identical across calls from
+    different processes on the same node at least until 
+    the next OS reboot.
+
+    The last resort method is uuid.getnode() that in some
+    rare OS configurations may return a random ID each time it is called.
+    However, this method should never be reached on a Linux system,
+    because reading from /proc/sys/kernel/random/boot_id will be tried prior
+    to that. If uuid.getnode() is reached, it will be called twice,
+    and exception raised if the values are not identical.
+
+    :param list extraIDFiles: Optional list of additional file names
+    to try reading node ID before trying default methods. ID
+    should be a single word (no spaces) on the first line of the file.
+
+    """
+    import os
+    idSourceFiles = extraIDFiles + ["/var/lib/dbus/machine-id",
+        "/proc/sys/kernel/random/boot_id"]
+    for idSourceFile in idSourceFiles:
+        if os.path.exists(idSourceFile):
+            try:
+                with open(idSourceFile,"r") as inp:
+                    nodeID = inp.readline().strip()
+            except EnvironmentError:
+                logger.warning(("Exception when trying to read ID file {}. "+\
+                        "Will try next method to get node ID").\
+                        format(idSourceFile), exc_info=True)
+            else:
+                if len(nodeID.split()) == 1:
+                    logger.debug("Obtained node ID {} from file {}".\
+                            format(nodeID,idSourceFile))
+                    break
+                else:
+                    logger.warning(("Node ID {} from file {} contains spaces."+\
+                            " Will try next method to get node ID").\
+                            format(nodeID,idSourceFile))
+    else:
+        import uuid
+        nodeIDs = []
+        for i_call in range(2):
+            nodeID = str(uuid.getnode()).strip()
+            assert len(nodeID.split())==1,("Node ID {} from uuid.getnode() "+\
+                    "contains spaces").format(nodeID)
+            nodeIDs.append(nodeID)
+        assert nodeIDs[0] == nodeIDs[1], ("Different node IDs {} received from"+\
```

Comment:
@cket OK, will do. However, this will also mean cryptic failures, or, worse, incorrect outputs, if they are using a shared file system for tmp, and the node ID is not determined by any of the techniques.